File: test/language/optional_members/python/OptionalRecursionTest.py
import zserio

import OptionalMembers


class OptionalRecursionTest(OptionalMembers.TestCase):
    def testConstructor(self):
        emptyBlock = self.api.Block(0)
        self.assertEqual(0, emptyBlock.byte_count)
        self.assertEqual(0, len(emptyBlock.data_bytes))
        self.assertEqual(0, emptyBlock.block_terminator)
        self.assertEqual(None, emptyBlock.next_data)

        emptyBlock = self.api.Block(0, [], 0, None)
        self.assertEqual(0, emptyBlock.byte_count)
        self.assertEqual(0, len(emptyBlock.data_bytes))
        self.assertEqual(0, emptyBlock.block_terminator)
        self.assertEqual(None, emptyBlock.next_data)

        emptyBlock = self.api.Block(byte_count_=0, data_bytes_=[], block_terminator_=0, next_data_=None)
        self.assertEqual(0, emptyBlock.byte_count)
        self.assertEqual(0, len(emptyBlock.data_bytes))
        self.assertEqual(0, emptyBlock.block_terminator)
        self.assertEqual(None, emptyBlock.next_data)

    def testEq(self):
        emptyBlock1 = self._createEmptyBlock()
        emptyBlock2 = self._createEmptyBlock()
        self.assertTrue(emptyBlock1 == emptyBlock2)

        block1 = self._createBlock(self.BLOCK1_DATA)
        self.assertFalse(block1 == emptyBlock1)

        block2 = self._createBlock(self.BLOCK1_DATA)
        self.assertTrue(block2 == block1)

        block12 = self._createBlock12(self.BLOCK1_DATA, self.BLOCK2_DATA)
        self.assertFalse(block12 == block1)

    def testHash(self):
        emptyBlock1 = self._createEmptyBlock()
        emptyBlock2 = self._createEmptyBlock()
        self.assertEqual(hash(emptyBlock1), hash(emptyBlock2))

        block1 = self._createBlock(self.BLOCK1_DATA)
        self.assertTrue(hash(block1) != hash(emptyBlock1))

        block2 = self._createBlock(self.BLOCK1_DATA)
        self.assertEqual(hash(block2), hash(block1))

        block12_1 = self._createBlock12(self.BLOCK1_DATA, self.BLOCK2_DATA)
        self.assertTrue(hash(block12_1) != hash(block1))

        block12_2 = self._createBlock12(self.BLOCK1_DATA, self.BLOCK2_DATA)
        self.assertEqual(hash(block12_1), hash(block12_2))

        block12_1.block_terminator = 0
        self.assertNotEqual(hash(block12_1), hash(block12_2))

        # use hardcoded values to check that the hash code is stable
        # using __hash__ to prevent 32-bit Python hash() truncation
        self.assertEqual(6240113, block12_1.__hash__())
        self.assertEqual(1846174533, block12_2.__hash__())

    def testIsNextDataSetAndUsed(self):
        block1 = self._createBlock(self.BLOCK1_DATA)
        self.assertFalse(block1.is_next_data_set())
        self.assertFalse(block1.is_next_data_used())

        block1.block_terminator = 1  # used but not set
        self.assertFalse(block1.is_next_data_set())
        self.assertTrue(block1.is_next_data_used())

        block12 = self._createBlock12(self.BLOCK1_DATA, self.BLOCK2_DATA)
        self.assertTrue(block12.is_next_data_set())
        self.assertTrue(block12.is_next_data_used())

        block12.block_terminator = 0  # set but not used
        self.assertTrue(block12.is_next_data_set())
        self.assertFalse(block12.is_next_data_used())

    def testResetNextData(self):
        block12 = self._createBlock12(self.BLOCK1_DATA, self.BLOCK2_DATA)
        self.assertTrue(block12.is_next_data_set())
        self.assertTrue(block12.is_next_data_used())

        block12.reset_next_data()  # used but not set
        self.assertFalse(block12.is_next_data_set())
        self.assertTrue(block12.is_next_data_used())
        self.assertEqual(None, block12.next_data)

    def testBitSizeOf(self):
        block1 = self._createBlock(self.BLOCK1_DATA)
        self.assertEqual(OptionalRecursionTest._getBlockBitSize(self.BLOCK1_DATA), block1.bitsizeof())

        block12 = self._createBlock12(self.BLOCK1_DATA, self.BLOCK2_DATA)
        self.assertEqual(
            OptionalRecursionTest._getBlock12BitSize(self.BLOCK1_DATA, self.BLOCK2_DATA), block12.bitsizeof()
        )

    def testInitializeOffsets(self):
        block1 = self._createBlock(self.BLOCK1_DATA)
        bitPosition = 1
        self.assertEqual(
            bitPosition + OptionalRecursionTest._getBlockBitSize(self.BLOCK1_DATA),
            block1.initialize_offsets(bitPosition),
        )

        block12 = self._createBlock12(self.BLOCK1_DATA, self.BLOCK2_DATA)
        self.assertEqual(
            bitPosition + OptionalRecursionTest._getBlock12BitSize(self.BLOCK1_DATA, self.BLOCK2_DATA),
            block12.initialize_offsets(bitPosition),
        )

    def testWriteBlock1(self):
        block1 = self._createBlock(self.BLOCK1_DATA)
        writer = zserio.BitStreamWriter()
        block1.write(writer)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        self._checkBlockInStream(reader, self.BLOCK1_DATA)

        reader.bitposition = 0
        readBlock1 = self.api.Block.from_reader(reader, len(self.BLOCK1_DATA))
        self.assertEqual(block1, readBlock1)

    def testWriteBlock12(self):
        block12 = self._createBlock12(self.BLOCK1_DATA, self.BLOCK2_DATA)
        writer = zserio.BitStreamWriter()
        block12.write(writer)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        self._checkBlock12InStream(reader, self.BLOCK1_DATA, self.BLOCK2_DATA)

        reader.bitposition = 0
        readBlock12 = self.api.Block.from_reader(reader, len(self.BLOCK1_DATA))
        self.assertEqual(block12, readBlock12)

    def _createEmptyBlock(self):
        return self.api.Block(0, None, 0, None)

    def _createBlock(self, blockData):
        return self.api.Block(len(blockData), blockData, 0, None)

    def _createBlock12(self, block1Data, block2Data):
        block2 = self._createBlock(block2Data)

        return self.api.Block(len(block1Data), block1Data, len(block2Data), block2)

    @staticmethod
    def _getBlockBitSize(blockData):
        return 8 * len(blockData) + 8

    @staticmethod
    def _getBlock12BitSize(block1Data, block2Data):
        return OptionalRecursionTest._getBlockBitSize(block1Data) + OptionalRecursionTest._getBlockBitSize(
            block2Data
        )

    def _checkBlockInStream(self, reader, blockData):
        for element in blockData:
            self.assertEqual(element, reader.read_bits(8))
        self.assertEqual(0, reader.read_bits(8))

    def _checkBlock12InStream(self, reader, block1Data, block2Data):
        for element in block1Data:
            self.assertEqual(element, reader.read_bits(8))
        self.assertEqual(len(block2Data), reader.read_bits(8))

        self._checkBlockInStream(reader, block2Data)

    BLOCK1_DATA = [1, 2, 3, 4, 5, 6]
    BLOCK2_DATA = [10, 9, 8, 7]

File: test/language/optional_members/cpp/OptionalRecursionTest.cpp
#include <array>

#include "gtest/gtest.h"
#include "optional_members/optional_recursion/Block.h"
#include "zserio/BitStreamReader.h"
#include "zserio/BitStreamWriter.h"
#include "zserio/RebindAlloc.h"

namespace optional_members
{
namespace optional_recursion
{

using allocator_type = Block::allocator_type;
template <typename T>
using vector_type = zserio::vector<T, allocator_type>;

class OptionalRecursionTest : public ::testing::Test
{
protected:
    template <size_t SIZE>
    void fillBlock(Block& block, const std::array<uint8_t, SIZE>& blockData)
    {
        vector_type<uint8_t> dataBytes;
        dataBytes.reserve(blockData.size());
        dataBytes.assign(blockData.begin(), blockData.end());
        block.setDataBytes(dataBytes);

        block.setBlockTerminator(0);

        block.initialize(static_cast<uint8_t>(blockData.size()));
    }

    template <size_t SIZE1, size_t SIZE2>
    void fillBlock(Block& block1, const std::array<uint8_t, SIZE1>& block1Data,
            const std::array<uint8_t, SIZE2>& block2Data)
    {
        Block block2;
        fillBlock(block2, block2Data);

        vector_type<uint8_t> dataBytes;
        dataBytes.reserve(block1Data.size());
        dataBytes.assign(block1Data.begin(), block1Data.end());
        block1.setDataBytes(dataBytes);

        block1.setBlockTerminator(static_cast<uint8_t>(block2Data.size()));
        block1.setNextData(block2);

        block1.initialize(static_cast<uint8_t>(block1Data.size()));
    }

    size_t getBlockBitSize(size_t blockDataSize)
    {
        return 8 * blockDataSize + 8;
    }

    size_t getBlockBitSize(size_t block1DataSize, size_t block2DataSize)
    {
        return getBlockBitSize(block1DataSize) + getBlockBitSize(block2DataSize);
    }

    template <size_t SIZE>
    void checkBlockInBitStream(zserio::BitStreamReader& reader, const std::array<uint8_t, SIZE>& blockData)
    {
        for (uint8_t data : blockData)
            ASSERT_EQ(data, reader.readBits(8));

        ASSERT_EQ(0, reader.readBits(8));
    }

    template <size_t SIZE1, size_t SIZE2>
    void checkBlockInBitStream(zserio::BitStreamReader& reader, const std::array<uint8_t, SIZE1>& block1Data,
            const std::array<uint8_t, SIZE2>& block2Data)
    {
        for (uint8_t data1 : block1Data)
            ASSERT_EQ(data1, reader.readBits(8));
        ASSERT_EQ(block2Data.size(), reader.readBits(8));

        checkBlockInBitStream(reader, block2Data);
    }

    static const std::array<uint8_t, 6> BLOCK1_DATA;
    static const std::array<uint8_t, 4> BLOCK2_DATA;

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
};

const std::array<uint8_t, 6> OptionalRecursionTest::BLOCK1_DATA = {1, 2, 3, 4, 5, 6};
const std::array<uint8_t, 4> OptionalRecursionTest::BLOCK2_DATA = {10, 9, 8, 7};

TEST_F(OptionalRecursionTest, isNextDataSetAndUsed)
{
    Block block1;
    fillBlock(block1, BLOCK1_DATA);
    ASSERT_FALSE(block1.isNextDataSet());
    ASSERT_FALSE(block1.isNextDataUsed());

    block1.setBlockTerminator(1); // used but not set
    ASSERT_FALSE(block1.isNextDataSet());
    ASSERT_TRUE(block1.isNextDataUsed());

    Block block12;
    fillBlock(block12, BLOCK1_DATA, BLOCK2_DATA);
    ASSERT_TRUE(block12.isNextDataSet());
    ASSERT_TRUE(block12.isNextDataUsed());

    block12.setBlockTerminator(0); // set but not used
    ASSERT_TRUE(block12.isNextDataSet());
    ASSERT_FALSE(block12.isNextDataUsed());
}

TEST_F(OptionalRecursionTest, resetNextData)
{
    Block block12;
    fillBlock(block12, BLOCK1_DATA, BLOCK2_DATA);
    ASSERT_TRUE(block12.isNextDataSet());
    ASSERT_TRUE(block12.isNextDataUsed());

    block12.resetNextData(); // used but not set
    ASSERT_FALSE(block12.isNextDataSet());
    ASSERT_TRUE(block12.isNextDataUsed());
    ASSERT_THROW(block12.getNextData(), zserio::CppRuntimeException);
}

TEST_F(OptionalRecursionTest, bitSizeOf)
{
    Block block1;
    fillBlock(block1, BLOCK1_DATA);
    ASSERT_EQ(getBlockBitSize(BLOCK1_DATA.size()), block1.bitSizeOf());

    Block block12;
    fillBlock(block12, BLOCK1_DATA, BLOCK2_DATA);
    ASSERT_EQ(getBlockBitSize(BLOCK1_DATA.size(), BLOCK2_DATA.size()), block12.bitSizeOf());

    block12.setBlockTerminator(0); // set but not used
    ASSERT_EQ(getBlockBitSize(BLOCK1_DATA.size()), block12.bitSizeOf());
}

TEST_F(OptionalRecursionTest, initializeOffsets)
{
    Block block1;
    fillBlock(block1, BLOCK1_DATA);
    const size_t bitPosition = 1;
    ASSERT_EQ(bitPosition + getBlockBitSize(BLOCK1_DATA.size()), block1.initializeOffsets(bitPosition));

    Block block12;
    fillBlock(block12, BLOCK1_DATA, BLOCK2_DATA);
    ASSERT_EQ(bitPosition + getBlockBitSize(BLOCK1_DATA.size(), BLOCK2_DATA.size()),
            block12.initializeOffsets(bitPosition));

    block12.setBlockTerminator(0); // set but not used
    ASSERT_EQ(bitPosition + getBlockBitSize(BLOCK1_DATA.size()), block12.initializeOffsets(bitPosition));
}

TEST_F(OptionalRecursionTest, operatorEquality)
{
    Block block1;
    fillBlock(block1, BLOCK1_DATA);

    Block block2;
    fillBlock(block2, BLOCK1_DATA);
    ASSERT_TRUE(block2 == block1);

    Block block12_1;
    fillBlock(block12_1, BLOCK1_DATA, BLOCK2_DATA);
    ASSERT_FALSE(block12_1 == block1);

    Block block12_2;
    fillBlock(block12_2, BLOCK1_DATA, BLOCK2_DATA);
    ASSERT_TRUE(block12_1 == block12_2);

    block12_1.setBlockTerminator(0); // set but not used
    ASSERT_FALSE(block12_1 == block12_2);

    block12_2.setBlockTerminator(0); // set but not used
    ASSERT_TRUE(block12_1 == block12_2);
}

TEST_F(OptionalRecursionTest, operatorLessThan)
{
    Block block1;
    fillBlock(block1, BLOCK1_DATA);
    Block block2;
    fillBlock(block2, BLOCK1_DATA);
    ASSERT_FALSE(block1 < block2);
    ASSERT_FALSE(block2 < block1);

    Block block12_1;
    fillBlock(block12_1, BLOCK1_DATA, BLOCK2_DATA);
    ASSERT_TRUE(block1 < block12_1);
    ASSERT_FALSE(block12_1 < block1);

    Block block12_2;
    fillBlock(block12_2, BLOCK1_DATA, BLOCK2_DATA);
    ASSERT_FALSE(block12_1 < block12_2);
    ASSERT_FALSE(block12_2 < block12_1);

    block12_1.setBlockTerminator(0); // set but not used
    ASSERT_TRUE(block12_1 < block12_2);
    ASSERT_FALSE(block12_2 < block12_1);

    block12_2.setBlockTerminator(0); // set but not used
    ASSERT_FALSE(block12_1 < block12_2);
    ASSERT_FALSE(block12_2 < block12_1);
}

TEST_F(OptionalRecursionTest, hashCode)
{
    Block block1;
    fillBlock(block1, BLOCK1_DATA);

    Block block2;
    fillBlock(block2, BLOCK1_DATA);
    ASSERT_EQ(block2.hashCode(), block1.hashCode());

    Block block12_1;
    fillBlock(block12_1, BLOCK1_DATA, BLOCK2_DATA);
    ASSERT_NE(block12_1.hashCode(), block1.hashCode());

    Block block12_2;
    fillBlock(block12_2, BLOCK1_DATA, BLOCK2_DATA);
    ASSERT_EQ(block12_1.hashCode(), block12_2.hashCode());

    block12_1.setBlockTerminator(0); // set but not used
    ASSERT_NE(block12_1.hashCode(), block12_2.hashCode());

    // use hardcoded values to check that the hash code is stable
    ASSERT_EQ(6240113, block12_1.hashCode());
    ASSERT_EQ(1846174533, block12_2.hashCode());

    block12_2.setBlockTerminator(0); // set but not used
    ASSERT_EQ(block12_1.hashCode(), block12_2.hashCode());
}

TEST_F(OptionalRecursionTest, writeBlock1)
{
    Block block1;
    fillBlock(block1, BLOCK1_DATA);

    zserio::BitStreamWriter writer(bitBuffer);
    block1.write(writer);
    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    checkBlockInBitStream(reader, BLOCK1_DATA);
    reader.setBitPosition(0);

    Block readBlock1(reader, BLOCK1_DATA.size());
    ASSERT_EQ(block1, readBlock1);
}

TEST_F(OptionalRecursionTest, writeBlock12)
{
    Block block12;
    fillBlock(block12, BLOCK1_DATA, BLOCK2_DATA);

    zserio::BitStreamWriter writer(bitBuffer);
    block12.write(writer);
    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    checkBlockInBitStream(reader, BLOCK1_DATA, BLOCK2_DATA);
    reader.setBitPosition(0);

    Block readBlock12(reader, BLOCK1_DATA.size());
    ASSERT_EQ(block12, readBlock12);

    block12.setBlockTerminator(0); // set but not used
    zserio::BitStreamWriter writer2(bitBuffer);
    block12.write(writer2);
    zserio::BitStreamReader reader2(writer2.getWriteBuffer(), writer2.getBitPosition(), zserio::BitsTag());
    checkBlockInBitStream(reader2, BLOCK1_DATA);
    reader2.setBitPosition(0);

    Block readBlock12_2(reader2, BLOCK1_DATA.size());
    ASSERT_EQ(block12, readBlock12_2);
}

} // namespace optional_recursion
} // namespace optional_members
